nemo_aligner/utils/trt_llm.py

```diff
@@ -55,32 +70,44 @@ def __init__(
         self.unload_engine_train = unload_engine_train
         self.trt_model_type = trt_model_type
         self.reshard_model = reshard_model
-        self.pad_id = tokenizer.eos_id
-
         self.trt_llm_exporter = TensorRTLLM(trt_model_dir, load_model=False)
         self._trtllm_model_compiled = False
 
         rng_generator = torch.Generator(device="cpu")
         seed = secrets.randbits(32) if seed is None else seed
         rng_generator.manual_seed(seed)
-
         self.rng_generator = rng_generator
 
+        self.pad_id = tokenizer.pad_id
+        end_id = tokenizer.eos_id
         end_strings = list(end_strings)
-        assert all(
-            "," not in end_string for end_string in end_strings
-        ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
 
-        end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+        # TRT-LLM uses different logic to compute the response length depending on if we specify stop_list or end_id
+        # we want to use the same logic(which is to include the stop token in response length) so we put end_id into the stop_list
+        # manually
+        if len(end_strings) == 0:
+            ids, offsets = [end_id], [1]
+        else:
+            assert all(
+                "," not in end_string for end_string in end_strings
+            ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
+
+            end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+            # use an arbitary ref string to obtain end_string ids
+            stop_list = tensorrt_llm_run.to_word_list_format(
+                end_strings, build_tokenizer(self.tokenizer), ref_str="green tea icecream"
+            )
+            # add the end_id, we only need to manipulate one slice
+            ids, offsets = stop_list[0].tolist()
+            ids = append_and_repad_list(ids, end_id, pad_id=0)
+            offsets = append_and_repad_list(offsets, max(offsets) + 1, pad_id=-1)
```

Comment:
Should we put these two lines under a
```python
if end_id not in ids:
```

Reply:
yeah sure dc0e9081bf2042dfb413ed91ba7e5ee38e67d5d0